File: lib/fluxsource/cwffluxsource.cc
#include "lib/core/globals.h"
#include "lib/data/fluxmap.h"
#include "lib/data/locations.h"
#include "lib/fluxsource/fluxsource.pb.h"
#include "lib/fluxsource/fluxsource.h"
#include "lib/external/catweasel.h"
#include "lib/config/proto.h"
#include <fstream>
#include <ranges>

struct CwfHeader
{
    char file_id[4];      // file ID - almost always "CWSF"
    uint8_t creator;      // usually just the CW_TYPE
    uint8_t file_type;    // indexed, etc.?
    uint8_t version;      // version of this file
    uint8_t clock_rate;   // clock rate used: 0, 1, 2 (2 = 28MHz)
    uint8_t drive_type;   // type of drive
    uint8_t cylinders;    // number of cylinders
    uint8_t heads;        // number of heads
    uint8_t index_mark;   // nonzero if index marks are included
    uint8_t step;         // track stepping interval
    uint8_t filler[15];   // reserved for expansion
    uint8_t comment[100]; // brief comment
};

struct CwfTrack
{
    uint8_t cylinder; // sequential
    uint8_t head;
    uint8_t unused[2];
    uint8_t length[4]; // little-endian
};

class CwfFluxSource : public TrivialFluxSource
{
public:
    CwfFluxSource(const CwfFluxSourceProto& config): _config(config)
    {
        _if.open(_config.filename(), std::ios::in | std::ios::binary);
        if (!_if.is_open())
            error("cannot open input file '{}': {}",
                _config.filename(),
                strerror(errno));

        _if.read((char*)&_header, sizeof(_header));
        check_for_error();

        if ((_header.file_id[0] != 'C') || (_header.file_id[1] != 'W') ||
            (_header.file_id[2] != 'S') || (_header.file_id[3] != 'F'))
            error("input not a CWF file");

        switch (_header.clock_rate)
        {
            case 1:
                _clockPeriod = 1e9 / 14161000.0;
                break;
            case 2:
                _clockPeriod = 1e9 / 28322000.0;
                break;
            default:
                error("unsupported clock rate");
        }

        std::cout << fmt::format("CWF {}x{} = {} cylinders, {} heads\n",
            _header.cylinders,
            _header.heads,
            _header.cylinders * _header.step,
            _header.heads);
        std::cout << fmt::format(
            "CWF sample clock rate: {} MHz\n", 1e3 / _clockPeriod);

        int cylinders = _header.cylinders * _header.heads;
        for (int i = 0; i < cylinders; i++)
        {
            CwfTrack trackHeader;
            _if.read((char*)&trackHeader, sizeof(trackHeader));
            check_for_error();

            uint32_t length =
                Bytes(trackHeader.length, 4).reader().read_le32() -
                sizeof(CwfTrack);
            unsigned track_number = trackHeader.cylinder * _header.step;

            off_t pos = _if.tellg();
            _trackOffsets[CylinderHead{track_number, trackHeader.head}] =
                std::make_pair(pos, length);
            _if.seekg(pos + length);
        }

        std::vector<CylinderHead> chs;
        for (auto& [key, value] : _trackOffsets)
            chs.push_back(key);
        _extraConfig.mutable_drive()->set_tracks(
            convertCylinderHeadsToString(chs));
    }

public:
    std::unique_ptr<const Fluxmap> readSingleFlux(int track, int side) override
    {
        const auto& p =
            _trackOffsets.find(CylinderHead{(unsigned)track, (unsigned)side});
        if (p == _trackOffsets.end())
            return std::make_unique<const Fluxmap>();

        off_t pos = p->second.first;
        size_t length = p->second.second;
        _if.seekg(pos);
        Bytes fluxdata(_if, length);
        check_for_error();

        return decodeCatweaselData(fluxdata, _clockPeriod);
    }

    void recalibrate() override {}

private:
    void check_for_error()
    {
        if (_if.fail())
            error("SCP read I/O error: {}", strerror(errno));
    }

private:
    const CwfFluxSourceProto& _config;
    std::ifstream _if;
    CwfHeader _header;
    nanoseconds_t _clockPeriod;
    std::map<CylinderHead, std::pair<off_t, size_t>> _trackOffsets;
};

std::unique_ptr<FluxSource> FluxSource::createCwfFluxSource(
    const CwfFluxSourceProto& config)
{
    return std::unique_ptr<FluxSource>(new CwfFluxSource(config));
}

File: java/com/cowlark/fluxengine/fluxsource/CwfFluxSource.java
package com.cowlark.fluxengine.fluxsource;

import com.cowlark.fluxengine.config.ConfigBuilder;
import com.cowlark.fluxengine.config.ConfigProto;
import com.cowlark.fluxengine.core.ByteReader;
import com.cowlark.fluxengine.core.Bytes;
import com.cowlark.fluxengine.core.FluxEngineException;
import com.cowlark.fluxengine.core.Logger;
import com.cowlark.fluxengine.external.Catweasel;
import com.cowlark.fluxengine.data.CylinderHead;
import com.cowlark.fluxengine.data.Fluxmap;
import com.cowlark.fluxengine.data.Locations;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/**
 * A flux source which reads a CWF flux file, ported from
 * lib/fluxsource/cwffluxsource.cc.
 */
public class CwfFluxSource extends TrivialFluxSource
{
    private final Bytes data;
    private final double clockPeriod;
    private final Map<CylinderHead, int[]> trackOffsets = new TreeMap<>();
    protected ConfigProto extraConfig;

    public CwfFluxSource(CwfFluxSourceProto config)
    {
        data = readFile(config.getFilename());

        ByteReader br = new ByteReader(data);
        Bytes fileId = br.read(4);
        if ((fileId.getByte(0) != 'C') || (fileId.getByte(1) != 'W') ||
                (fileId.getByte(2) != 'S') || (fileId.getByte(3) != 'F'))
            throw new FluxEngineException("input not a CWF file");

        br.read8(); /* creator */
        br.read8(); /* file_type */
        br.read8(); /* version */
        int clockRate = br.read8();
        br.read8(); /* drive_type */
        int cylinders = br.read8();
        int heads = br.read8();
        br.read8(); /* index_mark */
        int step = br.read8();
        br.skip(15); /* filler */
        br.skip(100); /* comment */

        switch (clockRate)
        {
            case 1:
                clockPeriod = 1e9 / 14161000.0;
                break;
            case 2:
                clockPeriod = 1e9 / 28322000.0;
                break;
            default:
                throw new FluxEngineException("unsupported clock rate");
        }

        Logger.logf("CWF %dx%d = %d cylinders, %d heads",
                cylinders,
                heads,
                cylinders * step,
                heads);
        Logger.logf("CWF sample clock rate: %d MHz", (int) (1e3 / clockPeriod));

        int numTracks = cylinders * heads;
        for (int i = 0; i < numTracks; i++)
        {
            int trackCylinder = br.read8();
            int trackHead = br.read8();
            br.skip(2); /* unused */
            int length = br.readLe32();
            int dataLength = length - 8;
            int pos = br.pos();
            trackOffsets.put(new CylinderHead(trackCylinder * step, trackHead),
                    new int[]{pos, dataLength});
            br.skip(dataLength);
        }

        List<CylinderHead> chs = new ArrayList<>(trackOffsets.keySet());
        ConfigProto.Builder builder = ConfigProto.newBuilder();
        builder.getDriveBuilder().setTracks(Locations.convertCylinderHeadsToString(chs));
        extraConfig = builder.build();
    }

    private static Bytes readFile(String filename)
    {
        try
        {
            return new Bytes(Files.readAllBytes(Path.of(filename)));
        } catch (IOException e)
        {
            throw new FluxEngineException(
                    "cannot open input file '" + filename + "': " + e.getMessage());
        }
    }

    @Override
    public void adjustConfig(ConfigBuilder configBuilder)
    {
        configBuilder.mergeConfig(extraConfig);
    }

    @Override
    public Fluxmap readSingleFlux(FluxReadParameters parameters)
    {
        int[] offset = trackOffsets.get(new CylinderHead(parameters.cylinder(), parameters.head()));
        if (offset == null)
            return new Fluxmap();

        Bytes fluxdata = data.slice(offset[0], offset[1]);
        return Catweasel.decodeCatweaselData(fluxdata, clockPeriod);
    }
}
